### ISSUE TITLE
Timeslice option on start or requestData method

### ISSUE THREAD

**Issue by pdesantis** [chrisguttandin/extendable-media-recorder#263]
During recording, I would love to be able to access the already-recorded data at regular intervals. My ultimate goal is to back up the data in case of a computer crash before the `mediaRecorder.stop()` method is called.

The native MediaRecorder provides a couple of options to help with this: the [timeslice parameter](https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/start) on its `start` method, and the [requestData method](https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/requestData).

Does this library support either of those options, or have another way to access data blobs before calling the `stop` method? If not, do you have any suggestions on how to go about implementing it? I'd be willing to try implementing it if it's not on your timeline.

Thanks!
Pat

**Comment by chrisguttandin** [comment]
Hi Pat,

unfortunately this is not yet implemented. Mainly because I don't really know how to do it. The spec says that "the individual Blobs need not be playable, but the combination of all the Blobs from a completed recording MUST be playable".

A WAV file needs to have the length of the file encoded in it's header. But once the first Blob is returned to the user there is no way to alternate the header anymore. Therefore the combination of all Blobs would be an invalid file.

Do you maybe know if there is a standard for creating WAV files of unknown length?

**Comment by pdesantis** [comment]
Ah, that's a good point. Unfortunately the closest way to support this that I've found is to just encode the maximum value (`0xFFFFFFFF`) and trust the player to handle it. [Source](https://stackoverflow.com/questions/25245439/writing-wav-files-of-unknown-length)

I think most players are smart enough to handle incorrect duration metadata. For example, MP3 files are [pretty notorious](https://stackoverflow.com/questions/20711488/html5-audio-tag-showing-wrong-duration-of-mp3-in-chrome) for having incorrect or missing duration metadata, but most players will play the audio just fine.

You could encode the correct header if `start` is called with no `timeslice`, and only encode duration `0xFFFFFFFF` if `timeslice` is set. It's certainly not ideal, but IMO is better than breaking away from the native MediaRecorder API.

**Comment by chrisguttandin** [comment]
Hi Pat,

thanks for looking that up. I found a [specification of the WAV format](http://www-mmsp.ece.mcgill.ca/Documents/AudioFormats/WAVE/WAVE.html) which implies that a WAV file could have more than one data chunk. I've never seen such a file but if that works each time a `dataavailable` event gets fired it could at least contain a valid data chunk. Only the overall file size would then be too large.

I will try to implement it like that.

**Comment by chrisguttandin** [comment]
Hi Pat, it's been a while. Many thanks for your patience. I tried a lot of things but in the end I implemented it as you suggested.

> You could encode the correct header if start is called with no timeslice, and only encode duration 0xFFFFFFFF if timeslice is set. It's certainly not ideal, but IMO is better than breaking away from the native MediaRecorder API.

The resulting files can be understood by FFmpeg and can be played by the VLC player and Quicktime. iTunes does display the wrong duration but still plays them.

I tried to create files with multiple data chunks but that didn't work. I then tried using wave list chunks (wavl) and/or fact chunks (fact) in all possible combinations. I just could not get it to work. I think using the biggest possible value as the file size is kind of a de facto standard. I also found [a hint in the source of FFmpeg](https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/wavdec.c#L429) that this is somehow expected. Other JavaScript packages like node-wav [do employ a similar technique](https://github.com/TooTallNate/node-wav/blob/master/lib/writer.js#L36) as well.

For now I only implemented the timeslice parameter. The requestData() method is still missing.
